### PR TITLE
Add some lowerings for assign statements.

### PR DESCRIPTION
**Pull request by ptersilie**
Once https://github.com/softdevteam/yk/pull/48 is merged, this should be good to go.

### CODE REVIEW

**Comment by ptersilie** [comment]
Oh, I didn't rebase against master. Let me do that first to get rid of all of those commits.

File: src/librustc/sir.rs [hunk 1]
```diff
@@ -102,7 +104,74 @@ impl SirFuncCx {
...
+        let lhs = ykpack::Local(lvalue.local.as_u32());
+        let rhs = self.lower_rvalue(rvalue);
+        ykpack::Statement::Assign(ykpack::Place::from(lhs), rhs)
+    }
+
+    fn lower_rvalue(&self, rvalue: &mir::Rvalue<'_>) -> ykpack::Rvalue {
+        match rvalue {
+            mir::Rvalue::Use(mir::Operand::Copy(place)) |
+            mir::Rvalue::Use(mir::Operand::Move(place)) => {
+                let local = ykpack::Local(place.local.as_u32());
```
Comment (vext01) [inline src/librustc/sir.rs]: Shall we make a `lower_local()` function so that if we ever diverge from having a 1-1 mapping, we do'nt have to look in a million places for where locals are created? It also gives us an opportunity to add a comment about the 1-1 mapping.

**Comment by vext01** [comment]
One small comment, otherwise all looks good.

File: src/librustc/sir.rs (hunk 1 above)
Comment (ptersilie) [reply src/librustc/sir.rs]: Fixed in 92d7e50. Will need squashing.

**Comment by ptersilie** [comment]
Fixed. Let me know when to squash.

File: src/librustc/sir.rs [hunk 2]
```diff
@@ -102,7 +104,74 @@ impl SirFuncCx {
...
+            mir::StatementKind::Assign(box (ref place, ref rvalue)) => {
+                let assign = self.lower_assign_stmt(place, rvalue);
+                self.push_stmt(bb, assign);
+            },
+            _ => self.push_stmt(bb, ykpack::Statement::Unimplemented(format!("{:?}", stmt)))
+        }
+    }
+
+    fn lower_assign_stmt(&self, lvalue: &mir::Place<'_>, rvalue: &mir::Rvalue<'_>) -> ykpack::Statement {
+        let lhs = ykpack::Local(lvalue.local.as_u32());
```
Comment (vext01) [inline src/librustc/sir.rs]: Shouldn't this be `lower_local()` now?

Comment (ptersilie) [reply src/librustc/sir.rs]: Oh I missed that one. Also, shouldn't we actually call it `lower_place`?

Comment (vext01) [reply src/librustc/sir.rs]: A place and a local are different concepts. Possible confusion?

Comment (ptersilie) [reply src/librustc/sir.rs]: Okay, then maybe I should pass in the local and not the place, e.g. `lower_local(place.local)`.

Comment (vext01) [reply src/librustc/sir.rs]: Yep. I'd expect `lower_local()` to accept a MIR local and return a SIR local. It shouldn't be to do with place at all.

Comment (ptersilie) [reply src/librustc/sir.rs]: Done.

**Comment by ptersilie** [comment]
Let me know when to squash.

File: src/librustc/sir.rs [hunk 3]
```diff
@@ -135,10 +135,10 @@ impl SirFuncCx {
         }
     }
 
-    fn lower_local(&self, place: &mir::Place<'_>) -> ykpack::Local {
+    fn lower_local(&self, local: &mir::Local) -> ykpack::Local {
```
Comment (vext01) [inline src/librustc/sir.rs]: Sorry, one more thing. There's no point in passing the local as a reference. It's small and implements `Copy`, so let's pass it by value.

File: src/librustc/sir.rs [hunk 4]
```diff
@@ -102,7 +104,79 @@ impl SirFuncCx {
...
+            mir::StatementKind::Assign(box (ref place, ref rvalue)) => {
+                let assign = self.lower_assign_stmt(place, rvalue);
+                self.push_stmt(bb, assign);
+            },
+            _ => self.push_stmt(bb, ykpack::Statement::Unimplemented(format!("{:?}", stmt)))
+        }
+    }
+
+    fn lower_assign_stmt(&self, lvalue: &mir::Place<'_>, rvalue: &mir::Rvalue<'_>) -> ykpack::Statement {
+        let lhs = self.lower_local(&lvalue.local);
```
Comment (bjorn3) [inline src/librustc/sir.rs]: Please either use a new function `lower_place`, or return `Statement::Unimplemented` when there is a projection. This kind of will implement later makes it hard to debug miscompilations later on. (I speak from experience)

File: src/librustc/sir.rs (hunk 3 above)
Comment (ptersilie) [reply src/librustc/sir.rs]: Doesn't `Local(x)` already make a copy of `x`?

Comment (vext01) [reply src/librustc/sir.rs]: Well yes, it does. But note that passing a reference copies a pointer address under the hood too!

I'm just trying to reduce syntax noise really. You can use a reference if you feel strongly.

Comment (ptersilie) [reply src/librustc/sir.rs]: Fair.

File: src/librustc/sir.rs (hunk 4 above)
Comment (ptersilie) [reply src/librustc/sir.rs]: I made a note so I won't forget about `projection` later, but adding `Unimplemented` now is probably a better idea. Fixed in 8ba020b

**Comment by ptersilie** [comment]
@vext01 Addressed both yours and @bjorn3's comments.

**Comment by vext01** [comment]
LGTM. Please squash.

**Comment by ptersilie** [comment]
Squashed!

**Comment by ptersilie** [comment]
Let's try again. Let me know when it's ok to squash.

**Comment by vext01** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed!

**Comment by vext01** [comment]
```
error[E0618]: expected function, found enum variant `ykpack::Rvalue::Unimplemented`
   --> src/librustc/sir.rs:135:27
    |
135 |                 Err(_) => ykpack::Rvalue::Unimplemented(format!("Use({:?})", place).to_string()),
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-----------------------------------------
    |                           |
    |                           call expression requires function
    |
help: `ykpack::Rvalue::Unimplemented` is a unit variant, you need to write it without the parenthesis
    |
135 |                 Err(_) => ykpack::Rvalue::Unimplemented,
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

CI cached a version of ykpack prior to your `Rvalue` change. That's why I thought we might need a 3-phase.

I'm just trying to think how we can fix this now.

**Comment by ptersilie** [comment]
As discussed offline, let's see if the problem was that I didn't commit the `Cargo.lock` file.

bors try

**Comment by vext01** [comment]
```
Dependencies not on the whitelist:
* maybe-uninit 
```

Looking at your Cargo.toml diff, the dep chain is: cargo -> crossbeam -> maybe_uninit.

But my lock file doesn't have `maybe-uninit` in there at all... Did you delete the lock file at some point?

I have:
```diff
diff --git a/Cargo.lock b/Cargo.lock
index bf1259d7ea4..5cffea70511 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -5524,7 +5524,7 @@ checksum = "e66366e18dc58b46801afbf2ca7661a9f59cc8c5962c29892b6039b4f86fa992"
 [[package]]
 name = "ykpack"
 version = "0.1.0"
-source = "git+https://github.com/softdevteam/yk#7b4cfe28cf0148813ee8efd9b58c3c023321eafd"
+source = "git+https://github.com/softdevteam/yk#141aa78d20a082f2518b56bc57181dbce84ce2f0"
 dependencies = [
  "fallible-iterator",
  "rmp-serde",
```

**Comment by ptersilie** [comment]
Yes, I deleted the lock file and rebuild it from scratch.

**Comment by vext01** [comment]
> Yes, I deleted the lock file and rebuild it from scratch.

That rarely works I'm afraid.

**Comment by ptersilie** [comment]
Okay, let me try again.

**Comment by ptersilie** [comment]
bors try

**Comment by vext01** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed!

**Comment by vext01** [comment]
The build succeeded, but github hasn't noticed...

**Comment by vext01** [comment]
Um, ignore that. I was looking at the wrong build.
